### PR TITLE
ARMA: consolidate P and Q

### LINKED ISSUE #784: ARMA: Pmax and Pmin not used
--------
Issue Description
--------
The ARMA inputs `Pmax` and `Pmin` are no longer useful, as they have to be the same value.  Instead, there should be a single input.

This is true for `Q` as well.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #784.

##### What are the significant changes in functionality due to this change request?
Changes the `Pmax` and `Pmin` to a single `P` parameter, and similarly for `Q`.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: doc/user_manual/rom.tex [hunk 1]
```diff
@@ -2922,8 +2922,6 @@ \subsubsection{ARMA}
 \end{equation*}
 where $x$ is a vector of dimension $n$, and $\phi_i$ and $\theta_j$ are both $n$ by $n$ matrices. When $q=0$, the above is
 autoregressive (AR); when $p=0$, the above is moving average (MA).
-While plans for optimizing $p$ and $q$ are under consideration, currently setting $Pmin$ and $Pmax$ to the
-same value is required, and similarly for $Qmin$ and $Qmax$.
 %The user is allowed to provide upper and lower limits for $p$
 %and $q$ (see below), and the training process will choose the optimal $p$ and $q$ that fall into the user-specified range.
```
Comment (wangcj05) [inline doc/user_manual/rom.tex]: These two lines are not needed, right?

File: tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml [hunk 2]
```diff
@@ -32,10 +32,8 @@
       <Target>Speed,Time</Target>
       <Features>scaling</Features>
       <pivotParameter>Time</pivotParameter>
-      <Pmax>2</Pmax>
-      <Pmin>2</Pmin>
-      <Qmax>1</Qmax>
-      <Qmin>1</Qmin>
+      <P>2</P>
+      <Q>1</Q>
       <outTruncation domain="positive">Speed</outTruncation>
```
Comment (wangcj05) [inline tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml]: update the TestInfo regarding these changes

File: tests/framework/ROM/TimeSeries/ARMA/arma_reseed_test.xml [hunk 3]
```diff
@@ -1,93 +1,92 @@
...
-        <Output>Speed,Time</Output>
-        <options>
-            <pivotParameter>Time</pivotParameter>
-        </options>
-      </HistorySet>
-    </DataObjects>
-</Simulation>
-
-
+      <revision author="alfoa" date="2018-04-26">Modified location of the test (ARMA folder)</revision>
```
Comment (wangcj05) [inline tests/framework/ROM/TimeSeries/ARMA/arma_reseed_test.xml]: update the TestInfo regarding this modifications.

File: tests/framework/ROM/TimeSeries/ARMA/multiple_arma.xml [hunk 4]
```diff
@@ -67,10 +67,8 @@
       <Target>A,B,Time</Target>
       <Features>scaling</Features>
       <pivotParameter>Time</pivotParameter>
-      <Pmax>2</Pmax>
-      <Pmin>2</Pmin>
-      <Qmax>1</Qmax>
-      <Qmin>1</Qmin>
+      <P>2</P>
+      <Q>1</Q>
```
Comment (wangcj05) [inline tests/framework/ROM/TimeSeries/ARMA/multiple_arma.xml]: TestInfo should be updated.

File: tests/framework/ROM/TimeSeries/ARMA/multiple_fourier.xml [hunk 5]
```diff
@@ -72,16 +71,21 @@
...
+      <!-- will apply to the rest (C,D) -->
       <FourierOrder>1,1</FourierOrder>
+      <P>3</P>
+      <Q>1</Q>
+      <P>3</P>
+      <Q>1</Q>
+      <P>3</P>
+      <Q>1</Q>
+      <P>3</P>
+      <Q>1</Q>
```
Comment (wangcj05) [inline tests/framework/ROM/TimeSeries/ARMA/multiple_fourier.xml]: 1. why P and Q and duplicated here
2. TestInfo Should be updated.

File: tests/framework/ROM/TimeSeries/ARMA/single_fourier.xml [hunk 6]
```diff
@@ -44,13 +44,17 @@
...
       <Fourier>86400</Fourier>
       <FourierOrder>2</FourierOrder>
+      <P>2</P>
+      <Q>1</Q>
+      <P>2</P>
+      <Q>1</Q>
+      <P>2</P>
+      <Q>1</Q>
+      <P>2</P>
+      <Q>1</Q>
```
Comment (wangcj05) [inline tests/framework/ROM/TimeSeries/ARMA/single_fourier.xml]: The same as above?

File: tests/framework/ROM/TimeSeries/ARMA/test_ARMA.xml [hunk 7]
```diff
@@ -51,13 +52,17 @@
...
       <Fourier>604800,86400</Fourier>
       <FourierOrder>2,2</FourierOrder>
+      <P>2</P>
+      <Q>1</Q>
+      <P>2</P>
+      <Q>1</Q>
+      <P>2</P>
+      <Q>1</Q>
+      <P>2</P>
+      <Q>1</Q>
```
Comment (wangcj05) [inline tests/framework/ROM/TimeSeries/ARMA/test_ARMA.xml]: See above comments

File: tests/framework/ROM/TimeSeries/ARMA/test_ARMA_parallel.xml [hunk 8]
```diff
@@ -56,13 +56,17 @@
...
       <Fourier>604800,86400</Fourier>
       <FourierOrder>2,2</FourierOrder>
+      <P>2</P>
+      <Q>1</Q>
+      <P>2</P>
+      <Q>1</Q>
+      <P>2</P>
+      <Q>1</Q>
+      <P>2</P>
+      <Q>1</Q>
```
Comment (wangcj05) [inline tests/framework/ROM/TimeSeries/ARMA/test_ARMA_parallel.xml]: see above comments

File: tests/framework/ROM/TimeSeries/ARMA/varma.xml [hunk 9]
```diff
@@ -109,10 +107,8 @@
       <seed>84</seed>
       <correlate>A,B,C</correlate>
       <pivotParameter>Time</pivotParameter>
-      <Pmax>3</Pmax>
-      <Pmin>3</Pmin>
-      <Qmax>1</Qmax>
-      <Qmin>1</Qmin>
+      <P>3</P>
+      <Q>1</Q>
```
Comment (wangcj05) [inline tests/framework/ROM/TimeSeries/ARMA/varma.xml]: update TestInfo

File: tests/framework/ROM/TimeSeries/ARMA/zero_filter.xml [hunk 10]
```diff
@@ -60,18 +60,17 @@
...
-      <Pmax>3</Pmax>
-      <Pmin>3</Pmin>
-      <Qmax>2</Qmax>
-      <Qmin>2</Qmin>
-      <Fourier>86400</Fourier> <!-- 1 year, 3 months, 1 day -->
+      <Fourier>86400</Fourier>
+      <!-- 1 year, 3 months, 1 day -->
       <FourierOrder>1</FourierOrder>
+      <P>3</P>
+      <Q>2</Q>
```
Comment (wangcj05) [inline tests/framework/ROM/TimeSeries/ARMA/zero_filter.xml]: update test info

File: doc/user_manual/rom.tex (hunk 1 above)
Comment (PaulTalbot-INL) [reply doc/user_manual/rom.tex]: It's likely to be added again in the future, but I can remove them for now.

File: tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml (hunk 2 above)
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/ARMA/arma_reseed_precursor.xml]: done

File: tests/framework/ROM/TimeSeries/ARMA/arma_reseed_test.xml (hunk 3 above)
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/ARMA/arma_reseed_test.xml]: Any changes here were just due to formatting in the conversion script; I think perhaps Windows line endings were removed.  Nothing was revised about the test, as far as I can tell.

File: tests/framework/ROM/TimeSeries/ARMA/multiple_arma.xml (hunk 4 above)
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/ARMA/multiple_arma.xml]: done

File: tests/framework/ROM/TimeSeries/ARMA/multiple_fourier.xml (hunk 5 above)
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/ARMA/multiple_fourier.xml]: fixed

File: tests/framework/ROM/TimeSeries/ARMA/single_fourier.xml (hunk 6 above)
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/ARMA/single_fourier.xml]: fixed

File: tests/framework/ROM/TimeSeries/ARMA/test_ARMA.xml (hunk 7 above)
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/ARMA/test_ARMA.xml]: fixed

File: tests/framework/ROM/TimeSeries/ARMA/test_ARMA_parallel.xml (hunk 8 above)
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/ARMA/test_ARMA_parallel.xml]: fixed

File: tests/framework/ROM/TimeSeries/ARMA/varma.xml (hunk 9 above)
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/ARMA/varma.xml]: added

File: tests/framework/ROM/TimeSeries/ARMA/zero_filter.xml (hunk 10 above)
Comment (PaulTalbot-INL) [reply tests/framework/ROM/TimeSeries/ARMA/zero_filter.xml]: fixed

**Comment by PaulTalbot-INL** [comment]
Conversion script has been added.

Draft email:

RAVEN ARMA users:

In recent changes to the ARMA model, the likelihood search for optimal lag parameters P and Q as been removed, and it was required for Pmax and Pmin to be the same value (similar to Q). As of RAVEN PR #785, we have removed Pmax and Pmin and replaced it with simply P, and similar for Q.

A conversion script has been written to help you make changes in your inputs, and is located in raven/scripts/conversionScripts/ARMA_pq.py.  It takes as a command-line argument the input you want to change; for example, "python ~/projects/raven/scripts/conversionScripts/ARMA_pq.py my_input.xml". It will create a backup file, "my_input.bak", for comparison.

For HYBRID users, this change has not been pushed to the RAVEN submodule of HYBRID yet, so no need to take action. We will notify you when this changes.

File: scripts/conversionScripts/ARMA_pq.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import xml.etree.ElementTree as ET
+import xml.dom.minidom as pxml
+import os
+
+def convert(tree,fileName=None):
+  """
+    Converts input files to be compatible with merge request #460:
```
Comment (wangcj05) [inline scripts/conversionScripts/ARMA_pq.py]: update the docstring, should be #785

File: scripts/conversionScripts/ARMA_pq.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+     - Removes "all" node
```
Comment (wangcj05) [inline scripts/conversionScripts/ARMA_pq.py]: update the explanation regarding this convert, ```all``` is for basic statistical?

File: scripts/conversionScripts/ARMA_pq.py (hunk 11 above)
Comment (PaulTalbot-INL) [reply scripts/conversionScripts/ARMA_pq.py]: Fixed docstrings, they were left from another conversion script.

**Comment by wangcj05** [comment]
PR looks good to me.

**Comment by PaulTalbot-INL** [comment]
It looks like the test timing out is repeatable on other Windows systems as well. Looking into options ...

**Comment by PaulTalbot-INL** [comment]
All tests are passing; the only changes aside from merging in `devel` is to update the segmented ARMA test to use P and Q instead of Pmax and Qmax.

**Comment by wangcj05** [comment]
checklists are satisfied.

**Comment by wangcj05** [comment]
This PR can be merged. @PaulTalbot-INL Could you email the users regarding the changes?

**Comment by wangcj05** [comment]
@PaulTalbot-INL merged.

**Pull request merged**